extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/HttpUpgradePolicy.java

```diff
@@ -0,0 +1,57 @@
+package io.quarkus.websockets.next;
+
+import io.quarkus.security.identity.SecurityIdentity;
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * A policy that controls which requests are allowed to upgrade the HTTP connection to a WebSocket connection.
+ * CDI beans implementing this interface are invoked on every request.
+ */
+public interface HttpUpgradePolicy {
+
+    /**
+     * This method inspects HTTP Upgrade context and either allows or denies upgrade to a WebSocket connection;
+     * CDI beans implementing this method must never use {@link RoutingContext} to send response directly.
+     *
+     * @param context {@link HttpUpgradeContext}
+     * @return check result; must never be null
+     */
+    Uni<CheckResult> checkUpgrade(HttpUpgradeContext context);
+
+    /**
+     * @param routingContext {@link RoutingContext}
+     * @param securityIdentity {@link SecurityIdentity}; the identity is null if the Quarkus Security extension is absent
+     */
+    record HttpUpgradeContext(RoutingContext routingContext, SecurityIdentity securityIdentity) {
+    }
+
+    final class CheckResult {
+
+        private static final CheckResult PERMIT_UPGRADE = new CheckResult(true, null);
+
+        private final boolean upgradePermitted;
+        private final int httpResponseCode;
```

Comment:
At some point we may need to add response headers that would be always added to the HTTP response, no matter if the upgrade was rejected or not.

Reply:
alright, I'll add it. It will require bit of a headers merging between policies, but I can make it effective.